lib/spack/llnl/util/filesystem.py

```diff
@@ -423,14 +423,36 @@ def mkdirp(*paths, **kwargs):
     Keyword Aguments:
         mode (permission bits or None, optional): optional permissions to
             set on the created directory -- use OS default if not provided
+        mode_intermediate (permission bits or None, optional):
+            same as mode, but for newly-created intermediate directories
     """
     mode = kwargs.get('mode', None)
+    mode_intermediate = kwargs.get('mode_intermediate', None)
     for path in paths:
         if not os.path.exists(path):
             try:
+                intermediate_folders = []
+                if mode_intermediate is not None:
+                    # detect missing intermediate folders
+                    intermediate_path = os.path.dirname(path)
+
+                    while intermediate_path:
+                        if os.path.exists(intermediate_path):
+                            break
+
+                        intermediate_folders.append(intermediate_path)
+                        intermediate_path = os.path.dirname(intermediate_path)
+
+                # create folders
                 os.makedirs(path)
+
+                # leaf folder permissions
                 if mode is not None:
                     os.chmod(path, mode)
+
+                # newly-created intermediate folders permissions
+                for intermediate_path in reversed(intermediate_folders):
```

Comment:
This needs an `if`-block, as we shouldn't chmod when `intermediate_mode is None`.

Reply:
The loop that changes permissions to intermediate folders is over `intermediate_folders`, that it is initialized to an empty list and it is populated (a few lines above) just `if mode_intermediate is not None` (and in case, it gets populated just by folders in the suffix that does not exists, leaf folder excluded).

Maybe I can try to improve code readability, but I think it does the correct thing.

Reply:
The loop that changes permissions to intermediate folders is over `intermediate_folders`, that it is initialized to an empty list and it is populated (a few lines above) just `if mode_intermediate is not None` (and in case, it gets populated just by folders in the suffix that does not exist, leaf folder excluded).

Maybe I can try to improve code readability, but I think it does the correct thing.

Reply:
Yeah you're right. Can you add a comment to that effect for readability?